tools/urtrace/collector.cpp

```diff
@@ -0,0 +1,305 @@
+#include "ur_params.hpp"
+#include "xpti/xpti_trace_framework.h"
+
+constexpr uint16_t TRACE_FN_BEGIN =
+    static_cast<uint16_t>(xpti::trace_point_type_t::function_with_args_begin);
+constexpr uint16_t TRACE_FN_END =
+    static_cast<uint16_t>(xpti::trace_point_type_t::function_with_args_end);
+constexpr std::string_view UR_STREAM_NAME = "ur";
+
+logger::Logger out = logger::create_logger("collector", true);
+
+#define ARGS_ENV "UR_COLLECTOR_ARGS"
+
+enum time_unit {
+    TIME_UNIT_AUTO,
+    TIME_UNIT_NS,
+    TIME_UNIT_US,
+    TIME_UNIT_MS,
+    TIME_UNIT_S,
+    MAX_TIME_UNIT,
+};
+
+const char *time_unit_str[MAX_TIME_UNIT] = {"auto", "ns", "us", "ms", "s"};
+
+std::string time_to_str(std::chrono::nanoseconds dur, enum time_unit unit) {
+    std::ostringstream ostr;
+
+    switch (unit) {
+    case TIME_UNIT_AUTO: {
+        if (dur.count() < 1000) {
+            return time_to_str(dur, TIME_UNIT_NS);
+        }
+        if (dur.count() < 1000 * 1000) {
+            return time_to_str(dur, TIME_UNIT_US);
+        }
+        if (dur.count() < 1000 * 1000 * 1000) {
+            return time_to_str(dur, TIME_UNIT_MS);
+        }
+        return time_to_str(dur, TIME_UNIT_S);
+    } break;
+    case TIME_UNIT_NS: {
+        ostr << dur.count() << "ns";
+    } break;
+    case TIME_UNIT_US: {
+        std::chrono::duration<double, std::micro> d = dur;
+        ostr << d.count() << "us";
+    } break;
+    case TIME_UNIT_MS: {
+        std::chrono::duration<double, std::milli> d = dur;
+        ostr << d.count() << "ms";
+    } break;
+    case TIME_UNIT_S: {
+        std::chrono::duration<double, std::ratio<1>> d = dur;
+        ostr << d.count() << "s";
+    } break;
+    default: {
+        out.error("invalid time unit {}", unit);
+        break;
+    }
+    }
+
+    return ostr.str();
+}
+
+/*
+ * Since this is a library that gets loaded alongside the traced program, it
+ * can't just accept arguments from the trace CLI tool directly. Instead, the
+ * arguments are passed through an environment variable. Users shouldn't set
+ * these directly.
+ *
+ * Available arguments (documented in trace cli tool):
+ * - "print_begin"
+ * - "profiling"
+ * - "time_unit:<auto,ns, ...>"
+ * - "filter:<regex>"
+ */
+static class cli_args {
+    std::optional<std::string>
+    arg_with_value(std::string_view name, const EnvVarMap::value_type &arg) {
+        if (arg.first != name) {
+            return std::nullopt;
+        }
+        if (arg.second.empty() || arg.second.size() > 1) {
+            out.warn("{} requires a single argument, skipping...", name);
+            return std::nullopt;
+        }
+        return arg.second.at(0);
+    }
+
+  public:
+    cli_args() {
+        print_begin = false;
+        profiling = false;
+        time_unit = TIME_UNIT_AUTO;
+        no_args = false;
+        filter = std::nullopt;
+        filter_str = std::nullopt;
+        if (auto args = getenv_to_map(ARGS_ENV, false)) {
+            for (auto arg : *args) {
```

Comment:
```
for (auto [arg_name, arg_value] : *args)
```

Reply:
Actually, I changed my mind :P It's more typing, but it does look a bit nicer.